packages/flutter_tools/lib/src/cache.dart

```diff
@@ -120,3 +132,91 @@ class MaterialFonts {
     });
   }
 }
+
+// TODO(devoncarew): Move the services download to here.
+// gs://flutter_infra/flutter/ed3014b3d337d025393bd894ffa2897e05d43e91/firebase/
+// gs://flutter_infra/flutter/ed3014b3d337d025393bd894ffa2897e05d43e91/gcm/
+
+class FlutterEngine {
+  FlutterEngine(this.cache);
+
+  static const String kName = 'engine';
+
+  final Cache cache;
+
+  List<String> _getEngineDirs() {
+    List<String> dirs = <String>['android-arm', 'android-x64'];
+
+    if (Platform.isMacOS)
+      dirs.addAll(<String>['ios', 'darwin-x64']);
+    else if (Platform.isLinux)
+      dirs.add('linux-x64');
+
+    return dirs;
+  }
+
+  List<String> _getPackageDirs() => <String>['sky_engine', 'sky_services'];
+
+  bool isUpToDate() {
+    Directory pkgDir = cache.getCacheDir('pkg');
+    for (String pkgName in _getPackageDirs()) {
+      Directory dir = new Directory(path.join(pkgDir.path, pkgName));
+      if (!dir.existsSync())
+        return false;
+    }
+
+    Directory engineDir = cache.getArtifactDirectory(kName);
+    for (String dirName in _getEngineDirs()) {
+      Directory dir = new Directory(path.join(engineDir.path, dirName));
+      if (!dir.existsSync())
+        return false;
+    }
+
+    return cache.getVersionFor(kName) == cache.getStampFor(kName);
+  }
+
+  Future<Null> download() async {
+    String engineVersion = cache.getVersionFor(kName);
+    String url = 'https://storage.googleapis.com/flutter_infra/flutter/$engineVersion/';
+
+    bool allDirty = engineVersion != cache.getStampFor(kName);
+
+    Directory pkgDir = cache.getCacheDir('pkg');
+    for (String pkgName in _getPackageDirs()) {
+      Directory dir = new Directory(path.join(pkgDir.path, pkgName));
+      if (!dir.existsSync() || allDirty) {
+        await _downloadItem('Downloading engine package $pkgName...',
+          url + pkgName + '.zip', dir);
+      }
+    }
+
+    Directory engineDir = cache.getArtifactDirectory(kName);
+    for (String dirName in _getEngineDirs()) {
+      Directory dir = new Directory(path.join(engineDir.path, dirName));
+      if (!dir.existsSync() || allDirty) {
+        await _downloadItem('Downloading engine artifacts $dirName...',
+          url + dirName + '/artifacts.zip', dir);
+        _makeFilesExecutable(dir);
+      }
+    }
+
+    cache.setStampFor(kName, cache.getVersionFor(kName));
+  }
+
+  void _makeFilesExecutable(Directory dir) {
+    for (FileSystemEntity entity in dir.listSync()) {
+      // chmod a+x "$ENGINE_ARTIFACT_PATH/darwin-x64/sky_snapshot" and sky_shell
+      if (entity is File && path.basename(entity.path).startsWith('sky_'))
```

Comment:
This is sort of magical.  We'll probably rename these at some point.  I guess it was pretty magical before too.